File: 7-event-sourcing/event-store-service/src/account/account.service.ts
/**
 * AccountService — xử lý commands và queries cho Event Store.
 *
 * Các luồng chính:
 *   1. appendEvent   — nhận command → validate → append event row vào Postgres.
 *   2. getProjection — đọc event log từ DB → replay → trả AccountState (read model).
 *   3. rebuildAll    — replay toàn bộ event log cho tất cả accounts → log kết quả.
 *   4. takeSnapshot  — tính AccountState hiện tại → lưu AccountSnapshot vào DB.
 *
 * (EN: AccountService — processes commands and queries for the Event Store.
 *
 * Main flows:
 *   1. appendEvent   — receive command → validate → append event row to Postgres.
 *   2. getProjection — load event log from DB → replay → return AccountState (read model).
 *   3. rebuildAll    — replay full event log for all accounts → log results.
 *   4. takeSnapshot  — compute current AccountState → persist AccountSnapshot.)
 */
import {
    BadRequestException,
    Injectable,
    Logger,
    NotFoundException,
} from "@nestjs/common"
import {
    InjectRepository,
} from "@nestjs/typeorm"
import {
    Repository,
    DataSource,
} from "typeorm"
import {
    AccountSnapshot,
    EventRecord,
} from "../entities"
import type {
    AccountDomainEvent,
} from "./events"
import type {
    CloseAccountCommand,
    DepositMoneyCommand,
    OpenAccountCommand,
    WithdrawMoneyCommand,
} from "./commands"
import {
    applyEvent,
    replayEvents,
} from "./handlers"
import type {
    AccountState,
} from "./handlers"

/** Mỗi bao nhiêu events thì tự động tạo snapshot. (EN: Auto-snapshot every N events.) */
const SNAPSHOT_THRESHOLD = 10

@Injectable()
export class AccountService {
    private readonly logger = new Logger(AccountService.name)

    constructor(
        @InjectRepository(EventRecord)
        private readonly eventRepo: Repository<EventRecord>,
        @InjectRepository(AccountSnapshot)
        private readonly snapshotRepo: Repository<AccountSnapshot>,
        private readonly dataSource: DataSource,
    ) {}

    // ─── Commands ────────────────────────────────────────────────────────────

    /**
     * Mở tài khoản mới — append event "AccountOpened".
     * Logic — tạo accountId random, append event version 1.
     * Code — `_appendEvent(accountId, "Account", "AccountOpened", payload)`.
     * (EN: Open a new account — append "AccountOpened" event.
     * Logic — generate random accountId, append event at version 1.
     * Code — `_appendEvent(accountId, "Account", "AccountOpened", payload)`.)
     */
    async openAccount(cmd: OpenAccountCommand): Promise<{ accountId: string }> {
        const accountId = `acc_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`
        const payload: AccountDomainEvent = {
            type: "AccountOpened",
            accountId,
            owner: cmd.owner,
            initialBalance: cmd.initialBalance,
        }
        await this._appendEvent(accountId, "Account", payload)
        this.logger.log(`AccountOpened: ${accountId} owner=${cmd.owner}`)
        return { accountId }
    }

    /**
     * Nạp tiền — append event "MoneyDeposited".
     * (EN: Deposit money — append "MoneyDeposited" event.)
     */
    async deposit(accountId: string, cmd: DepositMoneyCommand): Promise<AccountState> {
        const state = await this._loadState(accountId)
        if (state.status === "closed") {
            throw new BadRequestException(`Account ${accountId} is closed`)
        }
        const payload: AccountDomainEvent = {
            type: "MoneyDeposited",
            accountId,
            amount: cmd.amount,
        }
        await this._appendEvent(accountId, "Account", payload)
        return this._loadState(accountId)
    }

    /**
     * Rút tiền — validate số dư, append event "MoneyWithdrawn".
     * (EN: Withdraw money — validate balance, append "MoneyWithdrawn" event.)
     */
    async withdraw(accountId: string, cmd: WithdrawMoneyCommand): Promise<AccountState> {
        const state = await this._loadState(accountId)
        if (state.status === "closed") {
            throw new BadRequestException(`Account ${accountId} is closed`)
        }
        if (state.balance < cmd.amount) {
            throw new BadRequestException(
                `Insufficient balance: ${state.balance} < ${cmd.amount}`,
            )
        }
        const payload: AccountDomainEvent = {
            type: "MoneyWithdrawn",
            accountId,
            amount: cmd.amount,
        }
        await this._appendEvent(accountId, "Account", payload)
        return this._loadState(accountId)
    }

    /**
     * Đóng tài khoản — append event "AccountClosed".
     * (EN: Close account — append "AccountClosed" event.)
     */
    async closeAccount(accountId: string, cmd: CloseAccountCommand): Promise<AccountState> {
        const state = await this._loadState(accountId)
        if (state.status === "closed") {
            throw new BadRequestException(`Account ${accountId} is already closed`)
        }
        const payload: AccountDomainEvent = {
            type: "AccountClosed",
            accountId,
            reason: cmd.reason,
        }
        await this._appendEvent(accountId, "Account", payload)
        return this._loadState(accountId)
    }

    // ─── Queries ─────────────────────────────────────────────────────────────

    /**
     * Trả về AccountState hiện tại (projection) bằng cách replay event log.
     * Nếu có snapshot mới nhất, chỉ replay các event SAU snapshot đó.
     * (EN: Return current AccountState (projection) by replaying the event log.
     * If a snapshot exists, only replay events AFTER that snapshot version.)
     */
    async getProjection(accountId: string): Promise<AccountState> {
        return this._loadState(accountId)
    }

    /**
     * Trả về toàn bộ event log thô của một account (dùng cho time-travel / audit).
     * (EN: Return full raw event log for an account (for time-travel / audit).)
     */
    async getEventLog(accountId: string): Promise<EventRecord[]> {
        const events = await this.eventRepo.find({
            where: { aggregateId: accountId },
            order: { version: "ASC" },
        })
        if (events.length === 0) {
            throw new NotFoundException(`Account ${accountId} not found`)
        }
        return events
    }

    /**
     * Replay state của account tại một version cụ thể (time-travel).
     * (EN: Replay account state at a specific version (time-travel).)
     */
    async getStateAtVersion(accountId: string, version: number): Promise<AccountState> {
        const events = await this.eventRepo.find({
            where: { aggregateId: accountId },
            order: { version: "ASC" },
        })
        const sliced = events.filter(e => e.version <= version)
        if (sliced.length === 0) {
            throw new NotFoundException(
                `Account ${accountId} not found or no events at version ${version}`,
            )
        }
        return replayEvents(accountId, sliced)
    }

    /**
     * Rebuild tất cả projections — replay toàn bộ event log cho từng account, log kết quả.
     * Dùng để demo: "projection là derived data — có thể rebuild bất cứ lúc nào".
     * (EN: Rebuild all projections — replay full event log per account, log results.
     * Used to demo: "projection is derived data — can be rebuilt at any time".)
     */
    async rebuildAllProjections(): Promise<{ rebuilt: number; accounts: AccountState[] }> {
        // Lấy danh sách accountId duy nhất.
        // (EN: Fetch list of distinct accountIds.)
        const rows = await this.dataSource.query<Array<{ aggregate_id: string }>>(
            `SELECT DISTINCT aggregate_id FROM events WHERE aggregate_type = 'Account'`,
        )
        const accounts: AccountState[] = []
        for (const row of rows) {
            const id = row["aggregate_id"] ?? ""
            const state = await this._loadState(id)
            accounts.push(state)
            this.logger.log(
                `Rebuilt projection: ${id} balance=${state.balance} status=${state.status}`,
            )
        }
        return { rebuilt: accounts.length, accounts }
    }

    /**
     * Tạo snapshot cho account (lưu trạng thái hiện tại vào bảng account_snapshots).
     * (EN: Take a snapshot for an account (persist current state to account_snapshots table).)
     */
    async takeSnapshot(accountId: string): Promise<AccountSnapshot> {
        const state = await this._loadState(accountId)
        const existing = await this.snapshotRepo.findOne({
            where: { accountId, version: state.version },
        })
        if (existing) return existing

        const snap = this.snapshotRepo.create({
            accountId,
            version: state.version,
            state: state as unknown as Record<string, unknown>,
        })
        return this.snapshotRepo.save(snap)
    }

    // ─── Internals ───────────────────────────────────────────────────────────

    /**
     * Append một domain event vào bảng events trong một transaction.
     * Dùng SELECT ... FOR UPDATE để đảm bảo version monotonic, tránh xung đột concurrent.
     * (EN: Append one domain event to the events table inside a transaction.
     * Uses SELECT ... FOR UPDATE to guarantee monotonic version and prevent concurrent conflicts.)
     */
    private async _appendEvent(
        aggregateId: string,
        aggregateType: string,
        payload: AccountDomainEvent,
    ): Promise<EventRecord> {
        return this.dataSource.transaction(async manager => {
            // Dùng advisory lock theo hash của aggregateId để serialize concurrent writes cho cùng aggregate.
            // pg_advisory_xact_lock tự giải phóng khi transaction kết thúc.
            // (EN: Use advisory lock hashed from aggregateId to serialise concurrent writes for the same aggregate.
            // pg_advisory_xact_lock is automatically released when the transaction ends.)
            const lockKey = this._hashAggregateId(aggregateId)
            await manager.query(`SELECT pg_advisory_xact_lock($1)`, [lockKey])

            const [maxRow] = await manager.query<Array<{ max: string | null }>>(
                `SELECT MAX(version) AS max FROM events WHERE aggregate_id = $1`,
                [aggregateId],
            )
            const currentMax = maxRow?.["max"] != null ? Number(maxRow["max"]) : 0
            const nextVersion = currentMax + 1

            const record = manager.create(EventRecord, {
                aggregateId,
                aggregateType,
                version: nextVersion,
                eventType: payload.type,
                payload: payload as unknown as Record<string, unknown>,
            })

            const saved = await manager.save(EventRecord, record)

            // Auto-snapshot mỗi SNAPSHOT_THRESHOLD events.
            // (EN: Auto-snapshot every SNAPSHOT_THRESHOLD events.)
            if (nextVersion % SNAPSHOT_THRESHOLD === 0) {
                void this.takeSnapshot(aggregateId).catch(err =>
                    this.logger.warn(`Auto-snapshot failed for ${aggregateId}: ${String(err)}`),
                )
            }

            return saved
        })
    }

    /**
     * Hash aggregateId thành số nguyên 32-bit để dùng làm advisory lock key.
     * (EN: Hash aggregateId to a 32-bit integer for use as an advisory lock key.)
     */
    private _hashAggregateId(aggregateId: string): number {
        let hash = 0
        for (let i = 0; i < aggregateId.length; i++) {
            const chr = aggregateId.charCodeAt(i)
            hash = ((hash << 5) - hash) + chr
            hash |= 0 // Convert to 32-bit integer
        }
        return hash
    }

    /**
     * Load AccountState bằng cách:
     *   1. Tìm snapshot mới nhất (nếu có).
     *   2. Load các events có version > snapshot.version (hoặc toàn bộ nếu không có snapshot).
     *   3. Replay events lên trên snapshot state.
     * (EN: Load AccountState by:
     *   1. Finding latest snapshot (if any).
     *   2. Loading events with version > snapshot.version (or all if no snapshot).
     *   3. Replaying events on top of snapshot state.)
     */
    private async _loadState(accountId: string): Promise<AccountState> {
        // Tìm snapshot mới nhất.
        // (EN: Find latest snapshot.)
        const snapshot = await this.snapshotRepo.findOne({
            where: { accountId },
            order: { version: "DESC" },
        })

        // Load tất cả events sau snapshot (hoặc toàn bộ nếu không có snapshot).
        // (EN: Load all events after snapshot version (or all events if no snapshot).)
        const allEvents = await this.eventRepo.find({
            where: { aggregateId: accountId },
            order: { version: "ASC" },
        })

        if (allEvents.length === 0 && !snapshot) {
            throw new NotFoundException(`Account ${accountId} not found`)
        }

        if (snapshot) {
            // Bắt đầu từ snapshot state, fold thêm các events mới hơn.
            // (EN: Start from snapshot state, fold additional newer events on top.)
            const eventsAfterSnapshot = allEvents.filter(e => e.version > snapshot.version)
            let state = snapshot.state as unknown as AccountState
            for (const row of eventsAfterSnapshot) {
                state = applyEvent(
                    state,
                    row.payload as unknown as AccountDomainEvent,
                    row.occurredAt,
                    row.version,
                )
            }
            return state
        }

        // Không có snapshot — replay từ đầu.
        // (EN: No snapshot — replay from the beginning.)
        return replayEvents(accountId, allEvents)
    }
}

File: 7-event-sourcing/1-java/event-store-service/src/main/java/com/starci/eventstore/AccountService.java
package com.starci.eventstore;

import jakarta.persistence.EntityManager;
import jakarta.persistence.LockModeType;
import jakarta.persistence.PersistenceContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.server.ResponseStatusException;

import java.util.*;

/**
 * AccountService — processes commands and queries for the Event Store.
 *
 * <p>Main flows:
 * <ol>
 *   <li>appendEvent — receive command, validate, append event row to Postgres.</li>
 *   <li>getProjection — load event log from DB, replay, return AccountState.</li>
 *   <li>rebuildAll — replay full event log for all accounts, log results.</li>
 *   <li>takeSnapshot — compute current AccountState, persist AccountSnapshot.</li>
 * </ol>
 */
@Service
public class AccountService {

    private static final Logger log = LoggerFactory.getLogger(AccountService.class);

    /** Auto-snapshot every N events. */
    private static final int SNAPSHOT_THRESHOLD = 10;

    private final EventRecordRepository eventRepo;
    private final AccountSnapshotRepository snapshotRepo;

    @PersistenceContext
    private EntityManager em;

    public AccountService(EventRecordRepository eventRepo,
                          AccountSnapshotRepository snapshotRepo) {
        this.eventRepo = eventRepo;
        this.snapshotRepo = snapshotRepo;
    }

    // ─── Commands ────────────────────────────────────────────────────────────

    /**
     * Open a new account — append "AccountOpened" event.
     *
     * @param owner          account owner name
     * @param initialBalance initial balance (>= 0)
     * @return map with the new accountId
     */
    @Transactional
    public Map<String, String> openAccount(String owner, double initialBalance) {
        String accountId = "acc_" + System.currentTimeMillis() + "_"
                + Long.toHexString(new Random().nextLong() & 0xFFFFFFL);

        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("type", "AccountOpened");
        payload.put("accountId", accountId);
        payload.put("owner", owner);
        payload.put("initialBalance", initialBalance);

        appendEvent(accountId, "Account", "AccountOpened", payload);
        log.info("AccountOpened: {} owner={}", accountId, owner);
        return Map.of("accountId", accountId);
    }

    /**
     * Deposit money — append "MoneyDeposited" event.
     *
     * @param accountId target account
     * @param amount    deposit amount (> 0)
     * @return updated AccountState projection
     */
    @Transactional
    public AccountState deposit(String accountId, double amount) {
        AccountState state = loadState(accountId);
        if ("closed".equals(state.getStatus())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Account " + accountId + " is closed");
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("type", "MoneyDeposited");
        payload.put("accountId", accountId);
        payload.put("amount", amount);
        appendEvent(accountId, "Account", "MoneyDeposited", payload);
        return loadState(accountId);
    }

    /**
     * Withdraw money — validate balance, append "MoneyWithdrawn" event.
     *
     * @param accountId target account
     * @param amount    withdrawal amount (> 0)
     * @return updated AccountState projection
     */
    @Transactional
    public AccountState withdraw(String accountId, double amount) {
        AccountState state = loadState(accountId);
        if ("closed".equals(state.getStatus())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Account " + accountId + " is closed");
        }
        if (state.getBalance() < amount) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Insufficient balance: " + state.getBalance() + " < " + amount);
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("type", "MoneyWithdrawn");
        payload.put("accountId", accountId);
        payload.put("amount", amount);
        appendEvent(accountId, "Account", "MoneyWithdrawn", payload);
        return loadState(accountId);
    }

    /**
     * Close account — append "AccountClosed" event.
     *
     * @param accountId target account
     * @param reason    reason for closing
     * @return updated AccountState projection
     */
    @Transactional
    public AccountState closeAccount(String accountId, String reason) {
        AccountState state = loadState(accountId);
        if ("closed".equals(state.getStatus())) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST,
                    "Account " + accountId + " is already closed");
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("type", "AccountClosed");
        payload.put("accountId", accountId);
        payload.put("reason", reason);
        appendEvent(accountId, "Account", "AccountClosed", payload);
        return loadState(accountId);
    }

    // ─── Queries ─────────────────────────────────────────────────────────────

    /**
     * Return the current AccountState (projection) by replaying the event log.
     *
     * @param accountId target account
     * @return current projected state
     */
    @Transactional(readOnly = true)
    public AccountState getProjection(String accountId) {
        return loadState(accountId);
    }

    /**
     * Return the full raw event log for an account (audit trail / time-travel).
     *
     * @param accountId target account
     * @return list of event records in version order
     */
    @Transactional(readOnly = true)
    public List<EventRecord> getEventLog(String accountId) {
        List<EventRecord> events = eventRepo.findByAggregateIdOrderByVersionAsc(accountId);
        if (events.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
                    "Account " + accountId + " not found");
        }
        return events;
    }

    /**
     * Time-travel: return account state at a specific version.
     *
     * @param accountId target account
     * @param version   target version (inclusive)
     * @return account state at that version
     */
    @Transactional(readOnly = true)
    public AccountState getStateAtVersion(String accountId, int version) {
        List<EventRecord> events = eventRepo.findByAggregateIdOrderByVersionAsc(accountId);
        List<EventRecord> sliced = events.stream()
                .filter(e -> e.getVersion() <= version)
                .toList();
        if (sliced.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
                    "Account " + accountId + " not found or no events at version " + version);
        }
        return AccountProjection.replayEvents(accountId, sliced);
    }

    /**
     * Manually take a snapshot for the account.
     *
     * @param accountId target account
     * @return persisted snapshot
     */
    @Transactional
    public AccountSnapshot takeSnapshot(String accountId) {
        AccountState state = loadState(accountId);
        Optional<AccountSnapshot> existing =
                snapshotRepo.findByAccountIdAndVersion(accountId, state.getVersion());
        if (existing.isPresent()) return existing.get();

        Map<String, Object> stateMap = stateToMap(state);
        AccountSnapshot snap = new AccountSnapshot(accountId, state.getVersion(), stateMap);
        return snapshotRepo.save(snap);
    }

    /**
     * Rebuild all projections — replay full event log per account, log results.
     * Used to demonstrate: "projection is derived data — can be rebuilt at any time".
     *
     * @return result summary with rebuilt count and all account states
     */
    @Transactional(readOnly = true)
    public Map<String, Object> rebuildAllProjections() {
        List<String> accountIds = eventRepo.findDistinctAggregateIdsByType("Account");
        List<AccountState> accounts = new ArrayList<>();
        for (String id : accountIds) {
            AccountState state = loadState(id);
            accounts.add(state);
            log.info("Rebuilt projection: {} balance={} status={}",
                    id, state.getBalance(), state.getStatus());
        }
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("rebuilt", accounts.size());
        result.put("accounts", accounts);
        return result;
    }

    // ─── Internals ───────────────────────────────────────────────────────────

    /**
     * Append one domain event to the events table inside the current transaction.
     *
     * <p>Uses {@code pg_advisory_xact_lock} (via raw JPQL/native query) to serialise
     * concurrent writes for the same aggregate — mirrors the TS advisory lock pattern.
     *
     * @param aggregateId   aggregate ID
     * @param aggregateType aggregate type (e.g., "Account")
     * @param eventType     event type name
     * @param payload       event payload map
     * @return persisted EventRecord
     */
    private EventRecord appendEvent(String aggregateId, String aggregateType,
                                    String eventType, Map<String, Object> payload) {
        // Advisory lock — serialise concurrent writes for the same aggregate.
        long lockKey = hashAggregateId(aggregateId);
        em.createNativeQuery("SELECT pg_advisory_xact_lock(:key)")
                .setParameter("key", lockKey)
                .getSingleResult();

        // Compute next version.
        Number maxVersionNum = (Number) em.createQuery(
                "SELECT COALESCE(MAX(e.version), 0) FROM EventRecord e WHERE e.aggregateId = :id")
                .setParameter("id", aggregateId)
                .getSingleResult();
        int nextVersion = maxVersionNum.intValue() + 1;

        EventRecord record = new EventRecord(aggregateId, aggregateType,
                nextVersion, eventType, payload);
        em.persist(record);
        em.flush();

        // Auto-snapshot every SNAPSHOT_THRESHOLD events (best-effort, non-blocking).
        if (nextVersion % SNAPSHOT_THRESHOLD == 0) {
            try {
                takeSnapshot(aggregateId);
            } catch (Exception ex) {
                log.warn("Auto-snapshot failed for {}: {}", aggregateId, ex.getMessage());
            }
        }

        return record;
    }

    /**
     * Load AccountState by:
     * 1. Finding the latest snapshot (if any).
     * 2. Loading events with version > snapshot.version (or all if no snapshot).
     * 3. Replaying events on top of snapshot state.
     *
     * @param accountId target account
     * @return current projected state
     */
    private AccountState loadState(String accountId) {
        Optional<AccountSnapshot> snapshotOpt =
                snapshotRepo.findTopByAccountIdOrderByVersionDesc(accountId);
        List<EventRecord> allEvents =
                eventRepo.findByAggregateIdOrderByVersionAsc(accountId);

        if (allEvents.isEmpty() && snapshotOpt.isEmpty()) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND,
                    "Account " + accountId + " not found");
        }

        if (snapshotOpt.isPresent()) {
            AccountSnapshot snap = snapshotOpt.get();
            AccountState state = mapToState(snap.getState());
            List<EventRecord> eventsAfter = allEvents.stream()
                    .filter(e -> e.getVersion() > snap.getVersion())
                    .toList();
            for (EventRecord e : eventsAfter) {
                state = AccountProjection.applyEvent(state, e);
            }
            return state;
        }

        return AccountProjection.replayEvents(accountId, allEvents);
    }

    /**
     * Hash aggregateId to a 32-bit integer for use as pg_advisory_xact_lock key.
     * Uses the same djb2-style algorithm as the TypeScript source.
     *
     * @param aggregateId aggregate identifier
     * @return 32-bit hash as long
     */
    private long hashAggregateId(String aggregateId) {
        int hash = 0;
        for (char c : aggregateId.toCharArray()) {
            hash = ((hash << 5) - hash) + c;
        }
        return (long) hash;
    }

    // ─── State ↔ Map conversion ───────────────────────────────────────────────

    private Map<String, Object> stateToMap(AccountState s) {
        Map<String, Object> m = new LinkedHashMap<>();
        m.put("accountId", s.getAccountId());
        m.put("owner", s.getOwner());
        m.put("balance", s.getBalance());
        m.put("status", s.getStatus());
        m.put("version", s.getVersion());
        m.put("openedAt", s.getOpenedAt() != null ? s.getOpenedAt().toString() : null);
        m.put("lastEventAt", s.getLastEventAt() != null ? s.getLastEventAt().toString() : null);
        return m;
    }

    private AccountState mapToState(Map<String, Object> m) {
        AccountState s = new AccountState();
        s.setAccountId((String) m.get("accountId"));
        s.setOwner((String) m.get("owner"));
        s.setBalance(toDouble(m.get("balance")));
        s.setStatus((String) m.get("status"));
        s.setVersion(toInt(m.get("version")));
        Object openedAt = m.get("openedAt");
        if (openedAt != null) s.setOpenedAt(java.time.Instant.parse(openedAt.toString()));
        Object lastEventAt = m.get("lastEventAt");
        if (lastEventAt != null) s.setLastEventAt(java.time.Instant.parse(lastEventAt.toString()));
        return s;
    }

    private static double toDouble(Object v) {
        if (v instanceof Number n) return n.doubleValue();
        if (v == null) return 0;
        return Double.parseDouble(v.toString());
    }

    private static int toInt(Object v) {
        if (v instanceof Number n) return n.intValue();
        if (v == null) return 0;
        return Integer.parseInt(v.toString());
    }
}
